pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,313 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/pricing"
+	"github.com/aws/aws-sdk-go/service/pricing/pricingiface"
+	"github.com/samber/lo"
+	"knative.dev/pkg/logging"
+)
+
+// PricingProvider provides actual pricing data to the AWS cloud provider to allow it to make more informed decisions
+// regarding which instances to launch.  This is initialized at startup with a periodically updated static price list to
+// support running in locations where pricing data is unavailable.  In those cases the static pricing data provides a
+// relative ordering that is still more accurate than our previous pricing model.  In the event that a pricing update
+// fails, the previous pricing information is retained and used which may be the static initial pricing data if pricing
+// updates never succeed.
+type PricingProvider struct {
+	ec2     ec2iface.EC2API
+	pricing pricingiface.PricingAPI
+	region  string
+
+	mu                 sync.RWMutex
+	onDemandUpdateTime time.Time
+	onDemandPrices     map[string]float64
+	spotUpdateTime     time.Time
+	spotPrices         map[string]float64
+}
+
+// pricingUpdatePeriod is how often we try to update our pricing information after the initial update on startup
+const pricingUpdatePeriod = 12 * time.Hour
+
+// NewPricingAPI returns a pricing API configured based on a particular region
+func NewPricingAPI(sess *session.Session, region string) pricingiface.PricingAPI {
+	if sess == nil {
+		return nil
+	}
+	// pricing API doesn't have an endpoint in all regions
+	pricingAPIRegion := "us-east-1"
+	if strings.HasPrefix(region, "cn-") {
+		pricingAPIRegion = "cn-north-1"
+	}
+	return pricing.New(sess, &aws.Config{Region: aws.String(pricingAPIRegion)})
+}
+
+func NewPricingProvider(ctx context.Context, pricing pricingiface.PricingAPI, ec2Api ec2iface.EC2API, region string) *PricingProvider {
+	p := &PricingProvider{
+		region:             region,
+		onDemandUpdateTime: initialPriceUpdate,
+		onDemandPrices:     initialOnDemandPrices,
+		spotUpdateTime:     initialPriceUpdate,
+		spotPrices:         initialSpotPrices,
+		ec2:                ec2Api,
+		pricing:            pricing,
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("pricing"))
+
+	// perform an initial price update at startup
+	p.updatePricing(ctx)
```

Comment:
how long does this take? This blocks reconcilers from starting right?

Reply:
< 10 seconds, but no problem to move it to being parallel and run on the first iteration of the periodic loop. The downside there is you may provision the first time with stale pricing data.  WDYT?

Reply:
Do we know why our current startup time is so high? Ideally I'd like to get it down to a couple seconds, but if it's not possible, it may be worth being more correct.

Reply:
The pricing APIs aren't very performant.  I've got the price requests filtered as much as I can which helps, and I've just updated them to be in parallel. It takes < 3 seconds now.

```
{"level":"info","ts":1655781861.763466,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:142","msg":"updating pricing information"}
{"level":"info","ts":1655781862.709313,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:231","msg":"fetch pricing 945.664169ms took Compute Instance (bare metal)"}
{"level":"info","ts":1655781864.187744,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:382","msg":"updated spot pricing with 537 instance types"}
{"level":"info","ts":1655781864.187788,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:328","msg":"update spot pricing took 2.424180538s"}
{"level":"info","ts":1655781864.348746,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:231","msg":"fetch pricing 2.585108153s took Compute Instance"}
{"level":"info","ts":1655781864.348879,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:224","msg":"updated on-demand pricing with 545 instance types"}
{"level":"info","ts":1655781864.348892,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:171","msg":"update on-demand pricing took 2.585278096s"}
{"level":"info","ts":1655781864.348908,"logger":"fallback.aws.pricing","caller":"aws/pricing.go:145","msg":"update pricing took 2.585310603s"}
```

Reply:
I suspect if you're requesting pricing from Europe or Asia, it will take substantially longer since there's not a close endpoint